core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+
+                // 1. Run the sqlGetResultSchema command to get the result set
+                // metadata. Column metadata is sorted alphabetically by table
+                // and column name.
+                let get_result_schema_cmd =
+                    doc! {"sqlGetResultSchema": 1, "query": query, "schemaVersion": 1};
+
+                let get_result_schema_response: SqlGetResultSchemaResponse =
+                    bson::from_document(db.run_command(get_result_schema_cmd, None)?)
+                        .map_err(Error::BsonDeserialization)?;
+
+                let metadata = get_result_schema_response.process_metadata(current_db)?;
+
+                // 2. Run the $sql aggregation to get the result set cursor.
+                let pipeline = vec![doc! {"$sql": {
+                    "format": "odbc",
+                    "formatVersion": 1,
+                    "statement": query,
+                }}];
+
+                let options = query_timeout.map(|i| {
+                    AggregateOptions::builder()
+                        .max_time(Duration::from_millis(i as u64))
+                        .build()
+                });
+
+                let cursor = db.aggregate(pipeline, options)?;
+
+                Ok(MongoQuery {
+                    resultset_cursor: cursor,
+                    resultset_metadata: metadata,
+                    current: None,
+                })
+            }
+        }
     }
 
     // Return the number of fields/columns in the resultset
     fn _get_col_count(&self) -> u32 {
-        unimplemented!()
+        self.resultset_metadata.len() as u32
     }
 
     // Get the metadata for the column with the given index.
-    fn _get_col_metadata(&self, _col_index: u16) -> Result<MongoColMetadata> {
-        unimplemented!()
+    fn get_col_metadata(&self, col_index: u16) -> Result<&MongoColMetadata> {
+        self.resultset_metadata
+            .get(col_index as usize)
+            .map_or(Err(Error::ColIndexOutOfBounds(col_index)), |md| Ok(md))
     }
 }
 
 impl MongoStatement for MongoQuery {
     // Move the cursor to the next document and update the current row.
     // Return true if moving was successful, false otherwise.
+    // This method deserializes the current row and stores it in self.
     fn next(&mut self) -> Result<bool> {
-        unimplemented!()
+        let res = self.resultset_cursor.advance().map_err(Error::Mongo);
+        self.current = Some(self.resultset_cursor.deserialize_current()?);
+        res
     }
 
     // Get the BSON value for the cell at the given colIndex on the current row.
     // Fails if the first row as not been retrieved (next must be called at least once before getValue).
-    fn get_value(&self, _col_index: u16) -> Result<Option<Bson>> {
-        unimplemented!()
+    fn get_value(&self, col_index: u16) -> Result<Option<Bson>> {
+        let current = self.current.as_ref().ok_or(Error::InvalidCursorState)?;
```

Comment:
The spec says this is the error to return if someone attempts to get a value before advancing the cursor. That situation should be the only time we ever hit the `or` half of `ok_or`.